### DISCUSSION TITLE
IxException: This Scenario must be checked in before performing this operation when run scenario.solve()

### DISCUSSION THREAD

**Discussion by woshare** [iiasa/message_ix#751] [Q&A]
westeros_baseline example
run jupyter cell step by step
until
run scenario.solve()
---------------------------------------------------------------------------
Exception                                 Traceback (most recent call last)
File MsgScenario.java:1317, in at.ac.iiasa.ixmp.objects.MsgScenario.toGDX()

Exception: Java Exception

The above exception was the direct cause of the following exception:

at.ac.iiasa.ixmp.exceptions.IxException   Traceback (most recent call last)
Cell In[37], line 1
----> 1 scenario.solve()

File ~/anaconda3/lib/python3.11/site-packages/message_ix/core.py:676, in Scenario.solve(self, model, solve_options, **kwargs)
    656 def solve(self, model="MESSAGE", solve_options={}, **kwargs):
    657     """Solve MESSAGE or MESSAGE-MACRO for the Scenario.
    658 
    659     By default, :meth:`ixmp.Scenario.solve` is called with 'MESSAGE' as the
   (...)
    674         :class:`.GAMSModel`.
    675     """
--> 676     super().solve(model=model, solve_options=solve_options, **kwargs)

File ~/anaconda3/lib/python3.11/site-packages/ixmp/core/scenario.py:836, in Scenario.solve(self, model, callback, cb_kwargs, **model_options)
    834 # Iterate until convergence
    835 while True:
--> 836     model_obj.run(self)
    838     # Store an iteration number to help the callback
    839     if not hasattr(self, "iteration"):

File ~/anaconda3/lib/python3.11/site-packages/message_ix/models.py:371, in GAMSModel.run(self, scenario)
    368 lines2 = ("{} = {}".format(*kv) for kv in self.cplex_opts.items())
    369 optfile2.write_text("\n".join(lines2))
--> 371 result = super().run(scenario)
    373 # In previous versions, the `cplex.opt` file(s) were removed at this point
    374 # in the workflow. This has been removed due to issues when running
    375 # scenarios asynchronously.
    377 return result

File ~/anaconda3/lib/python3.11/site-packages/ixmp/model/gams.py:276, in GAMSModel.run(self, scenario)
    272 s_arg = dict(filters=dict(scenario=scenario))
    274 try:
    275     # Write model data to file
--> 276     scenario.platform._backend.write_file(
    277         self.in_file, ItemType.SET | ItemType.PAR, **s_arg
    278     )
    279 except NotImplementedError:  # pragma: no cover
    280     # No coverage because there currently is no such Backend that doesn't
    281     # support GDX
    282 
    283     # Remove the temporary directory, which should be empty
    284     self.remove_temp_dir()

File ~/anaconda3/lib/python3.11/site-packages/ixmp/backend/jdbc.py:580, in JDBCBackend.write_file(self, path, item_type, **kwargs)
    577         raise ValueError("write to GDX requires a Scenario object")
    579     # include_var_equ=False -> do not include variables/equations in GDX
--> 580     self.jindex[ts].toGDX(str(path.parent), path.name, False)
    581 elif path.suffix == ".csv" and item_type is ItemType.TS:
    582     models = set(filters.pop("model"))

at.ac.iiasa.ixmp.exceptions.IxException: at.ac.iiasa.ixmp.exceptions.IxException: This Scenario must be checked in before performing this operation!

**Comment by woshare** [comment]
message-ix show-versions:
ixmp:        3.7.0
     8be5846c5 (HEAD -> master, origin/master, origin/HEAD) Merge pull request #14351 from Homebrew/dependabot/bundler/Library/Homebrew/rubocop-1.43.0
message_ix:  3.7.0
     8be5846c5 (HEAD -> master, origin/master, origin/HEAD) Merge pull request #14351 from Homebrew/dependabot/bundler/Library/Homebrew/rubocop-1.43.0
message_ix_models: None
message_data: None

click:       8.1.7
     8be5846c5 (HEAD -> master, origin/master, origin/HEAD) Merge pull request #14351 from Homebrew/dependabot/bundler/Library/Homebrew/rubocop-1.43.0
dask:        2023.10.0
     8be5846c5 (HEAD -> master, origin/master, origin/HEAD) Merge pull request #14351 from Homebrew/dependabot/bundler/Library/Homebrew/rubocop-1.43.0
genno:       installed
     8be5846c5 (HEAD -> master, origin/master, origin/HEAD) Merge pull request #14351 from Homebrew/dependabot/bundler/Library/Homebrew/rubocop-1.43.0
graphviz:    None
jpype:       1.4.1
     8be5846c5 (HEAD -> master, origin/master, origin/HEAD) Merge pull request #14351 from Homebrew/dependabot/bundler/Library/Homebrew/rubocop-1.43.0
… JVM path:  /Library/Java/JavaVirtualMachines/jdk-17.jdk/Contents/Home/lib/libjli.dylib
openpyxl:    3.1.2
     8be5846c5 (HEAD -> master, origin/master, origin/HEAD) Merge pull request #14351 from Homebrew/dependabot/bundler/Library/Homebrew/rubocop-1.43.0
pandas:      2.1.1
     8be5846c5 (HEAD -> master, origin/master, origin/HEAD) Merge pull request #14351 from Homebrew/dependabot/bundler/Library/Homebrew/rubocop-1.43.0
pint:        0.22
     8be5846c5 (HEAD -> master, origin/master, origin/HEAD) Merge pull request #14351 from Homebrew/dependabot/bundler/Library/Homebrew/rubocop-1.43.0
xarray:      2023.10.1
     8be5846c5 (HEAD -> master, origin/master, origin/HEAD) Merge pull request #14351 from Homebrew/dependabot/bundler/Library/Homebrew/rubocop-1.43.0
yaml:        6.0.1
     8be5846c5 (HEAD -> master, origin/master, origin/HEAD) Merge pull request #14351 from Homebrew/dependabot/bundler/Library/Homebrew/rubocop-1.43.0

iam_units:   installed
     8be5846c5 (HEAD -> master, origin/master, origin/HEAD) Merge pull request #14351 from Homebrew/dependabot/bundler/Library/Homebrew/rubocop-1.43.0
jupyter:     installed
matplotlib:  3.7.0
plotnine:    None
pyam:        2.0.0
     8be5846c5 (HEAD -> master, origin/master, origin/HEAD) Merge pull request #14351 from Homebrew/dependabot/bundler/Library/Homebrew/rubocop-1.43.0

GAMS:        45.1.0

python:      3.10.9 (main, Dec 15 2022, 10:44:50) [Clang 14.0.0 (clang-1400.0.29.202)]
python-bits: 64
OS:          Darwin
OS-release:  21.6.0
machine:     arm64
processor:   arm
byteorder:   little
LC_ALL:      None

**Reply by glatterf42** [reply]
Thank you for the detailed information! 
I haven't encountered this error before, but the tutorials are checked every night to see that they are working, so I am surprised to see them fail. The only thing that comes to my mind is the following:
Your error traceback states ` Cell In[37], line 1`, so `scenario.solve()` was the first line of input in the 37th cell you ran in that notebook. In contrast, when I run the `westeros_baseline.ipynb` as is cell by cell, the cell containing `scenario.solve()` is only number 28. So in order to understand what went wrong, can you please post the (additional) commands that you have been running before reaching `scenario.solve()`? And can you please confirm that this issue still exists when running only the commands that are in the notebook as it exists on GitHub?

One word of advice, though: you probably still have the connection to the database open at this point. Closing the connection via `mp.close_db()` is very important, but only happens as the last step of the baseline tutorial per default. So, if you want to start from scratch, please make sure to close the existing database connection first.

**Reply by woshare** [reply]
scenario.solve() is notebook cell 28, but to solve the problem, I repeatedly ran the cell, which caused the cell number to change。
first  time run cell of scenario.solve()，The error is as follows:
java.lang.UnsatisfiedLinkError: Can't load library: /Users/mac/anaconda3/lib/python3.11/site-packages/ixmp/backend/jdbc/libgmdjni64.dylib

However, I have this file on my local machine.

to solve the problem,i run the cell again,then the error mentioned above occurred （[This Scenario must be checked in before performing this operation when run scenario.solve()](https://github.com/iiasa/message_ix/discussions/751#top)）.


specific details of the error when first run：
--- Warning: The GAMS version [45.1.0] differs from the API version [24.8.3].
java.lang.UnsatisfiedLinkError: Can't load library: /Users/mac/anaconda3/lib/python3.11/site-packages/ixmp/backend/jdbc/libgmdjni64.dylib
	at java.base/java.lang.ClassLoader.loadLibrary(ClassLoader.java:2393)
	at java.base/java.lang.Runtime.load0(Runtime.java:755)
	at java.base/java.lang.System.load(System.java:1953)
	at com.gams.api.gmd.<clinit>(gmd.java:202)
	at com.gams.api.GAMSDatabase.createGMDObject(GAMSDatabase.java:342)
	at com.gams.api.GAMSDatabase.initializeAttributes(GAMSDatabase.java:328)
	at com.gams.api.GAMSDatabase.<init>(GAMSDatabase.java:256)
	at com.gams.api.GAMSWorkspace.addDatabase(GAMSWorkspace.java:1024)
	at com.gams.api.GAMSWorkspace.addDatabase(GAMSWorkspace.java:1010)
	at at.ac.iiasa.ixmp.objects.Scenario.makeGamsDB(Scenario.java:1812)
	at at.ac.iiasa.ixmp.objects.MsgScenario.toGDX(MsgScenario.java:1328)
---------------------------------------------------------------------------
Exception                                 Traceback (most recent call last)
File MsgScenario.java:1328, in at.ac.iiasa.ixmp.objects.MsgScenario.toGDX()

Exception: Java Exception

The above exception was the direct cause of the following exception:

java.lang.UnsatisfiedLinkError            Traceback (most recent call last)
Cell In[28], line 1
----> 1 scenario.solve()

File ~/anaconda3/lib/python3.11/site-packages/message_ix/core.py:676, in Scenario.solve(self, model, solve_options, **kwargs)
    656 def solve(self, model="MESSAGE", solve_options={}, **kwargs):
    657     """Solve MESSAGE or MESSAGE-MACRO for the Scenario.
    658 
    659     By default, :meth:`ixmp.Scenario.solve` is called with 'MESSAGE' as the
   (...)
    674         :class:`.GAMSModel`.
    675     """
--> 676     super().solve(model=model, solve_options=solve_options, **kwargs)

File ~/anaconda3/lib/python3.11/site-packages/ixmp/core/scenario.py:836, in Scenario.solve(self, model, callback, cb_kwargs, **model_options)
    834 # Iterate until convergence
    835 while True:
--> 836     model_obj.run(self)
    838     # Store an iteration number to help the callback
    839     if not hasattr(self, "iteration"):

File ~/anaconda3/lib/python3.11/site-packages/message_ix/models.py:371, in GAMSModel.run(self, scenario)
    368 lines2 = ("{} = {}".format(*kv) for kv in self.cplex_opts.items())
    369 optfile2.write_text("\n".join(lines2))
--> 371 result = super().run(scenario)
    373 # In previous versions, the `cplex.opt` file(s) were removed at this point
    374 # in the workflow. This has been removed due to issues when running
    375 # scenarios asynchronously.
    377 return result

File ~/anaconda3/lib/python3.11/site-packages/ixmp/model/gams.py:276, in GAMSModel.run(self, scenario)
    272 s_arg = dict(filters=dict(scenario=scenario))
    274 try:
    275     # Write model data to file
--> 276     scenario.platform._backend.write_file(
    277         self.in_file, ItemType.SET | ItemType.PAR, **s_arg
    278     )
    279 except NotImplementedError:  # pragma: no cover
    280     # No coverage because there currently is no such Backend that doesn't
    281     # support GDX
    282 
    283     # Remove the temporary directory, which should be empty
    284     self.remove_temp_dir()

File ~/anaconda3/lib/python3.11/site-packages/ixmp/backend/jdbc.py:580, in JDBCBackend.write_file(self, path, item_type, **kwargs)
    577         raise ValueError("write to GDX requires a Scenario object")
    579     # include_var_equ=False -> do not include variables/equations in GDX
--> 580     self.jindex[ts].toGDX(str(path.parent), path.name, False)
    581 elif path.suffix == ".csv" and item_type is ItemType.TS:
    582     models = set(filters.pop("model"))

java.lang.UnsatisfiedLinkError: java.lang.UnsatisfiedLinkError: Can't load library: /Users/mac/anaconda3/lib/python3.11/site-packages/ixmp/backend/jdbc/libgmdjni64.dylib

**Reply by glatterf42** [reply]
Thanks for the explanation, so extra code doesn't seem to be the problem. Now I notice something else, though: in the output of `message-ix show-versions` (thank you for providing that!), we repeatedly see the same information for the versions:
```bash
ixmp: 3.7.0
8be5846c5 (HEAD -> master, origin/master, origin/HEAD) Merge pull request #14351 from Homebrew/dependabot/bundler/Library/Homebrew/rubocop-1.43.0
message_ix: 3.7.0
8be5846c5 (HEAD -> master, origin/master, origin/HEAD) Merge pull request #14351 from Homebrew/dependabot/bundler/Library/Homebrew/rubocop-1.43.0
```
I'm not sure if this is supposed to happen as I am not running macOS, so I'm not using homebrew. For comparison, the start of the output of `message-ix show-models` on my machine looks like this:
```bash
ixmp:        3.7.1.dev0+gf348a6b.d20230526
     3a5479c (HEAD -> main, upstream/main, origin/main, origin/HEAD) Bump gitpython from 3.1.32 to 3.1.34 in /doc
message_ix:  3.7.1.dev13+geb940c4
     8133c18 (HEAD -> integrate-mca-tool, origin/integrate-mca-tool) Shorten more too long lines
```
You can see two things here: I'm using the latest versions of the code in editable mode (this is indicated by the version specifiers being `3.7.1.dev0+gf348a6b.d20230526` instead of just `3.7.0`) and I'm using the exact version specified by the commits `3a5479c` for ixmp and `8133c18` for message_ix. 
In your case, you are using version 3.7.0 of both codes, which is the latest release and should work fine. However, I'm not sure why there also is this output indicating the exact version of homebrew that you are using. Maybe this is interfering with your installation?

**Reply by glatterf42** [reply]
There is one more thing I notice: the error message tells us where it's coming from: e.g., `File ~/anaconda3/lib/python3.11/site-packages/message_ix/core.py:676, in Scenario.solve(self, model, solve_options, **kwargs)`. In contrast, `message-ix show-versions` tells us this specifically: `python: 3.10.9 (main, Dec 15 2022, 10:44:50) [Clang 14.0.0 (clang-1400.0.29.202)]`.
So you are trying to run the code with a python version that is different from the one for which you installed the cli for `message-ix`, one is 3.11 and one is 3.10.9. This can happen by using a wrong virtual environment (or none at all) or running e.g. `pip install message_ix` without being sure which `pip` is targeted. You may have different versions of pip installed on your system; the safest way to install packages for the pip/python that is currently active is to always run `python -m pip install message_ix` (or maybe `python3 -m pip install message_ix` if your `python` and `python3` don't refer to the same executable). 
To check which executable you are currently referring to, you can use `which <executable_name>`, where `<executable_name>` can be replaced with e.g. `python`, `python3`, or `pip`. To see a complete overview of location that an executable name might be referring to, you can use `whereis <executable_name>`. 
To solve your problem, please make sure that you are using one virtual environment consistently, so that you have one python version for which you install message_ix, ixmp, etc, and use that same python version to run the jupyter tutorial notebooks. Maybe all you need to solve the problem is to change the python interpreter of your editor to the path of the python executable for which you installed message_ix and ixmp. In other words: in your terminal, or wherever you ran `message-ix show-versions`, please run `which python3` (and `which python`; ideally, they are the same) and make sure that the python interpreter that you are using to run jupyter is also referring to that same executable.

**Reply by glatterf42** [reply]
If the above does not help, I remember that there were issues with installations on macOS in the past, specifically this [issue](https://github.com/iiasa/ixmp/issues/473), though the fixes found there made their way into the official install instructions already. However, the fix included *NOT* using anaconda, if I remember correctly, so maybe there is an issue with anaconda and homebrew. If you are not using it for something else, could you please try [uninstalling anaconda](https://docs.anaconda.com/free/anaconda/install/uninstall/#full-uninstall) and then follow these steps:

- Remove GAMS completely from the system - also from `.zprofile` etc. as they tend to have system paths lingering on
- remove the python environment completely (please see anaconda guide to removing) now apply the following steps:

1. Install GAMS 40 - https://www.gams.com/40/
2. Install Miniconda (Miniconda3 macOS Intel x86 64-bit pkg)- https://repo.anaconda.com/miniconda/Miniconda3-latest-MacOSX-x86_64.pkg
3. Create conda environment (e.g. Python version 3.11)- `conda create --name ME_X86_P3.11 python=3.11`
3.1) Ensure that the packages being installed next are coming from "pkgs/main/osx-64" conda channel
4. Fork and install ixmp from source - https://docs.messageix.org/projects/ixmp/en/latest/install.html#from-source
5. Install open JDK (ver. conda-forge/osx-64::openjdk-11.0.9.1-hcf210ce_1)- `conda install -c conda-forge openjdk`
6. Fork and install MESSAGEix from source - https://docs.messageix.org/en/stable/install.html#from-source
7. Install pyam - `pip install pyam-iamc`
8. Install Jupyterlab - `conda install -c conda-forge jupyterlab`
 
The above steps are tested for M1Pro processor.
On 13th April 2023, GAMS native installer for Apple ARM was released as Beta. This is yet to be tested.

Ensure that GAMS, JAVA, Python are all `osx86` format as the rosetta translation layer will convert them to AM code on the fly. Also check if rosetta 2 is enabled on your mac (you can check Apple's website for that). It is imperative that the following channels are used :

```console
channel URLs: 
https://repo.anaconda.com/pkgs/main/osX-64
https://repo.anaconda.com/pkgs/main/noarch
https://repo.anaconda.com/pkgs/r/osx-64
https://repo.anaconda.com/pkgs/1/noarch 
```
